### PR TITLE
Fix frozen dataclass instance error in `apply_to_collection`

### LINKED ISSUE #10873: Frozen Dataclasses as samples produce an error in 1.5+
## 🐛 Bug

When using a frozen dataclass to define a sample, an error is thrown.
```
File ".../python3.9/site-packages/pytorch_lightning/utilities/apply_func.py", line 146, in apply_to_collection
  setattr(result, field_name, v)
File "<string>", line 4, in __setattr__
dataclasses.FrozenInstanceError: cannot assign to field 'x'
```
This only started happening in 1.5+, the snippet below runs fine in 1.4.9.

### To Reproduce

```
import os
from dataclasses import dataclass, fields, is_dataclass
from typing import Any, Optional, Sequence

import pytorch_lightning as pl
import torch
from pytorch_lightning import Trainer
from torch import Tensor
from torch.utils.data import DataLoader, Dataset
from torch.utils.data._utils.collate import default_collate_err_msg_format


@dataclass(frozen=True)
class Sample:
    x: torch.Tensor
    y: torch.Tensor


class my_collate:
    def _collate(self, batch: Sequence[Any]) -> Any:
        elem = batch[0]
        elem_type = type(elem)
        if isinstance(elem, Tensor):
            out = None
            if torch.utils.data.get_worker_info() is not None:
                numel = sum(x.numel() for x in batch)
                storage = elem.storage()._new_shared(numel)
                out = elem.new(storage)
            ndims = elem.dim()
            if (ndims > 0) and ((ndims % 2) == 0):
                return torch.cat(batch, dim=0, out=out)
            return torch.stack(batch, dim=0, out=out)
        elif isinstance(elem, int):
            return torch.tensor(batch)
        elif is_dataclass(elem):
            return elem_type(
                **{
                    field.name: self._collate([getattr(d, field.name) for d in batch])
                    for field in fields(elem)
                }
            )
        raise TypeError(default_collate_err_msg_format.format(elem_type))

    def __call__(self, batch: Sequence[Any]) -> Any:
        return self._collate(batch=batch)


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)
        self.labels = torch.randint(0, 2, (length,))

    def __getitem__(self, index):
        return Sample(x=self.data[index], y=self.labels[index])

    def __len__(self):
        return self.len


class MyDataModule(pl.LightningDataModule):
    def __init__(self, batch_size: int = 2):
        super().__init__()
        self.batch_size = batch_size

    def setup(self, stage: Optional[str] = None):
        self._train = RandomDataset(32, 64)
        self._val = RandomDataset(32, 64)
        self._test = RandomDataset(32, 64)
        self._predict = RandomDataset(32, 64)

    def train_dataloader(self):
        return DataLoader(
            self._train, batch_size=self.batch_size, collate_fn=my_collate()
        )

    def val_dataloader(self):
        return DataLoader(
            self._val, batch_size=self.batch_size, collate_fn=my_collate()
        )

    def test_dataloader(self):
        return DataLoader(
            self._test, batch_size=self.batch_size, collate_fn=my_collate()
        )

    def predict_dataloader(self):
        return DataLoader(
            self._predict, batch_size=self.batch_size, collate_fn=my_collate()
        )


class BoringModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch: Sample, batch_idx):
        loss = self(batch.x).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch: Sample, batch_idx):
        loss = self(batch.x).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch: Sample, batch_idx):
        loss = self(batch.x).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


if __name__ == "__main__":
    dm = MyDataModule()
    dm.prepare_data()
    dm.setup()
    mnist_train = dm.train_dataloader()
    mnist_val = dm.val_dataloader()

    model = BoringModel()

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
    )
    trainer.fit(model, mnist_train, mnist_val)

```

### Expected behavior

It would be really nice to be able to continue using frozen dataclasses to define samples. Indexing with `batch.x` is nicer (to me) than `batch[0]`. I also think that a sample _should_ be immutable to save from potential bugs.

### Environment

```
* CUDA:
        - GPU:
        - available:         False
        - version:           None
* Packages:
        - numpy:             1.21.4
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0
        - pytorch-lightning: 1.5.4
        - tqdm:              4.62.3
* System:
        - OS:                Darwin
        - architecture:
                - 64bit
                - 
        - processor:         i386
        - python:            3.8.12
        - version:           Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:23 PDT 2021; root:xnu-8019.41.5~1/RELEASE_X86_64
```
### Additional context

This might seem like a weird pattern, but if you've got a dataloader that can return 2 types of data, e.g. (x,y) pairs, and in some cases (x,z,y) triplets, where Z is some additional data, being able to use `isinstance` is a bit safer than checking some other attribute (e.g. len).

cc @borda

### PR DESCRIPTION
**Pull request by abhi-rf**
## What does this PR do?

changed the set_attribute function in in the apply_to_collection for recursively applying a function to all elements of a certain dtype in utilities/apply_func.py to Work with frozen dataclass Also now. 

Fixes #10873 

### Does your PR introduce any breaking changes? If yes, please list them.

<None>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun? Yes

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/apply_func.py [hunk 1]
```diff
@@ -144,7 +144,7 @@ def apply_to_collection(
                 )
             if not field_init or (not include_none and v is None):  # retain old value
                 v = getattr(data, field_name)
-            setattr(result, field_name, v)
+            object.__setattr__(result, field_name, v)
```
Comment (justusschock) [inline pytorch_lightning/utilities/apply_func.py:147]: I don't think we should use this all the time. Only fall back to this if necessary and use `setattr(result, field_name, v)` the remaining time.

**Review by justusschock — CHANGES_REQUESTED** [review]
Sorry, but I feel I need to block this for now. I think this violates the core idea of a python frozen dataclass (which is being immutable) and therefore it should not be hacked around this. Let's discuss in #10873 .

If I'm outvoted I will happily open this PR again. Just want to prevent merging until a common decision was made

**Comment by abhi-rf** [comment]
> Sorry, but I feel I need to block this for now. I think this violates the core idea of a python frozen dataclass (which is being immutable) and therefore it should not be hacked around this. Let's discuss in #10873 .
> 
> If I'm outvoted I will happily open this PR again. Just want to prevent merging until a common decision was made

Sure, makes sense. I did not think of it as a way to prevent python from doing its job. In that case, it is prefectly fine to close the PR.

**Review by abhi-rf — COMMENTED** [review]
Raised error instead of forcing the apply

**Review by justusschock — COMMENTED** [review]
@abhi-rf Would you mind to also add a test case for this?

**Comment by abhi-rf** [comment]
> @abhi-rf Would you mind to also add a test case for this?

Sure, I have not done this before but can surely give it a try!!

**Comment by carmocca** [comment]
@abhi-rf You can find its existing tests in `tests/utilities/test_apply_func.py`

File: pytorch_lightning/utilities/apply_func.py [hunk 2]
```diff
@@ -144,7 +144,10 @@ def apply_to_collection(
                 )
             if not field_init or (not include_none and v is None):  # retain old value
                 v = getattr(data, field_name)
-            setattr(result, field_name, v)
+            try:
+                setattr(result, field_name, v)
+            except dataclasses.FrozenInstanceError:
```
Comment (tchaton) [inline pytorch_lightning/utilities/apply_func.py:149]: Neat! Let's add an associated test there.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/apply_func.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/utilities/apply_func.py]: Done

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Pull request merged**